core/src/main/scala/org/http4s/internal/parsing/Rfc3986.scala

```diff
@@ -14,21 +14,45 @@
  * limitations under the License.
  */
 
-package org.http4s.internal.parsing
+package org.http4s
+package internal.parsing
 
 import cats.parse.Parser.{char, charIn, string1}
-import cats.parse.{Parser, Parser1}
-import cats.parse.Rfc5234.{digit, hexdig}
-import org.http4s.Uri.{Ipv4Address, Ipv6Address}
-import cats.syntax.applicative._
+import cats.parse.Parser1
+import cats.syntax.all._
 
 /** Common rules defined in Rfc3986
   *
   * @see [[https://tools.ietf.org/html/rfc3986]]
   */
 private[http4s] object Rfc3986 {
+  def alpha: Parser1[Char] = Rfc2234.alpha
 
-  private def ipv4Internal[A](f: (Byte, Byte, Byte, Byte) => A): Parser1[A] = {
+  def digit: Parser1[Char] = Rfc2234.digit
+
+  /* The spec references RFC2234, which is 0-9A-F, but it also
+   * explicitly permits lowercase. */
+  val hexdig: Parser1[Char] =
+    digit.orElse1(charIn("ABCDEF"))
+
+  /* unreserved    = ALPHA / DIGIT / "-" / "." / "_" / "~" */
+  val unreserved: Parser1[Char] =
+    alpha.orElse1(digit).orElse1(charIn("-._~"))
+
+  /* pct-encoded   = "%" HEXDIG HEXDIG */
+  val pctEncoded: Parser1[Char] =
+    (char('%') *> digit ~ digit).map { case (a, b) => (((a - '0') >> 4) + (b - '0')).toChar }
```

Comment:
Decoding this per-char is likely to be suboptimal, and maybe we should discourage it by returning `Parser1[Unit]`.